CRM/Contribute/BAO/Contribution.php

```diff
@@ -1775,7 +1787,19 @@ public static function transitionComponents($params, $processContributionObject
     elseif ($contributionStatusId == array_search('Failed', $contributionStatuses)) {
       if (is_array($memberships)) {
         foreach ($memberships as $membership) {
-          if ($membership) {
+          $update = TRUE;
+          //Update Membership status if there is no other completed contribution associated with the membership.
+          $relatedContributions = CRM_Member_BAO_Membership::getMembershipContributionId($membership->id, TRUE);
+          foreach ($relatedContributions as $contriId) {
+            if ($contriId == $contributionId) {
+              continue;
+            }
+            $status = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $contriId, 'contribution_status_id');
+            if ($status == array_search('Completed', $contributionStatuses)) {
```

Comment:
and here - probably most readable to resolve the name at the start of the function where both contributionStatusId & previousStatusId are resolved. It seems like we might need to make sure previous status is always resolved right at the start